File: ios/SafeBox/Disguise/Numpad/NumpadEntry.swift
import CoreGraphics

/// The PIN pad's display-only entry model (decisions §2.3). Pure so the dot
/// rules are unit-testable without a view.
///
/// The count is NOT the recorder: the host owns the buffer, and this only
/// decides how many dots to draw. Presses beyond the visible cap still animate,
/// buzz and emit a token — they simply add no dot.
struct NumpadEntry: Equatable, Sendable {
    /// Same value as `TokenRecorder.maxTokens`; beyond it the row stops growing.
    static let visibleDotCap = 32
    static let baseDotSize: CGFloat = 12
    static let baseDotGap: CGFloat = 12
    static let minDotSize: CGFloat = 6

    private(set) var count = 0

    mutating func append() { count += 1 }

    mutating func removeLast() {
        guard count > 0 else { return }
        count -= 1
    }

    mutating func clear() { count = 0 }

    var visibleDots: Int { min(count, Self.visibleDotCap) }

    /// Dot size and gap shrink uniformly when the row would exceed the column,
    /// with a floor of 6.
    static func dotMetrics(columnWidth: CGFloat, dots: Int) -> (size: CGFloat, gap: CGFloat) {
        guard dots > 1, columnWidth > 0 else { return (baseDotSize, baseDotGap) }
        let natural = CGFloat(dots) * baseDotSize + CGFloat(dots - 1) * baseDotGap
        guard natural > columnWidth else { return (baseDotSize, baseDotGap) }
        let scale = columnWidth / natural
        return (max(minDotSize, baseDotSize * scale), max(minDotSize, baseDotGap * scale))
    }
}

/// PIN pad geometry (decisions §8).
enum NumpadMetrics {
    static let columnMaxWidth: CGFloat = 320
    static let keyGap: CGFloat = 24
    static let minKeyDiameter: CGFloat = 64
    static let maxKeyDiameter: CGFloat = 80
    static let dotsToKeypadGap: CGFloat = 40

    static func sideMargin(width: CGFloat) -> CGFloat {
        min(max(width * 0.04, 12), 20)
    }

    static func columnWidth(width: CGFloat) -> CGFloat {
        min(width - 2 * sideMargin(width: width), columnMaxWidth)
    }

    static func keyDiameter(columnWidth: CGFloat) -> CGFloat {
        min(max((columnWidth - 2 * keyGap) / 3, minKeyDiameter), maxKeyDiameter)
    }
}

File: android/app/src/main/java/com/calcplus/calculator/feature/numpad/NumpadEntry.kt
package com.calcplus.calculator.feature.numpad

import com.calcplus.calculator.core.disguise.DisguiseMode
import com.calcplus.calculator.core.lock.TokenRecorder

/**
 * The PIN pad's display rules, kept pure so they are unit-testable without a
 * composition (decisions §2.3). This counts *dots*, nothing else: the face has
 * no buffer, no token list and no idea what the digits were — the host's
 * [TokenRecorder] owns all of that.
 */
object NumpadEntry {
    /** Layout constants (decisions §8). */
    const val COLUMN_MAX_WIDTH_DP = 320f
    const val KEY_GAP_DP = 24f
    const val KEY_MIN_DIAMETER_DP = 64f
    const val KEY_MAX_DIAMETER_DP = 80f
    const val DOTS_TO_KEYPAD_GAP_DP = 40f
    const val DOT_DIAMETER_DP = 12f
    const val DOT_GAP_DP = 12f
    const val DOT_SHRINK_FLOOR_DP = 6f

    /** Presses beyond this still animate, buzz and emit — they just add no dot. */
    const val VISIBLE_DOT_CAP = TokenRecorder.MAX_TOKENS

    fun afterDigit(count: Int): Int = count + 1

    /** Backspace on an empty entry is a no-op, exactly like the recorder's. */
    fun afterBackspace(count: Int): Int = (count - 1).coerceAtLeast(0)

    fun afterClear(): Int = 0

    /**
     * §2.2 rule 3: the capture modes clear the entry immediately on commit (the
     * caption tells the outcome); `disguise` and `verifyCurrent` keep the dots
     * on screen and keep accepting input until the host tears the face down or
     * pulses it.
     */
    fun afterCommit(count: Int, mode: DisguiseMode): Int = when (mode) {
        DisguiseMode.CAPTURE_NEW, DisguiseMode.CONFIRM_NEW -> 0
        DisguiseMode.DISGUISE, DisguiseMode.VERIFY_CURRENT -> count
    }

    fun visibleDots(count: Int): Int = count.coerceIn(0, VISIBLE_DOT_CAP)

    /** Circular key diameter for a column of [columnWidthDp] (3 keys, 2 gaps). */
    fun keyDiameter(columnWidthDp: Float): Float =
        ((columnWidthDp - 2f * KEY_GAP_DP) / 3f)
            .coerceIn(KEY_MIN_DIAMETER_DP, KEY_MAX_DIAMETER_DP)

    /**
     * Dot diameter and gap for [dots] dots in [columnWidthDp]. Both shrink
     * uniformly when the row would overflow, with a floor of
     * [DOT_SHRINK_FLOOR_DP].
     */
    fun dotMetrics(dots: Int, columnWidthDp: Float): Pair<Float, Float> {
        if (dots <= 0) return DOT_DIAMETER_DP to DOT_GAP_DP
        val natural = dots * DOT_DIAMETER_DP + (dots - 1) * DOT_GAP_DP
        if (natural <= columnWidthDp) return DOT_DIAMETER_DP to DOT_GAP_DP
        val scale = (columnWidthDp / natural).coerceIn(0f, 1f)
        val size = (DOT_DIAMETER_DP * scale).coerceAtLeast(DOT_SHRINK_FLOOR_DP)
        val gap = (DOT_GAP_DP * scale).coerceAtLeast(DOT_SHRINK_FLOOR_DP / 2f)
        return size to gap
    }
}
